packages/compile-solidity/compileWithPragmaAnalysis.js

```diff
@@ -0,0 +1,139 @@
+const CompilerSupplier = require("./compilerSupplier");
+const Config = require("@truffle/config");
+const semver = require("semver");
+const Profiler = require("./profiler");
+const fse = require("fs-extra");
+const { run } = require("./run");
+const OS = require("os");
+const cloneDeep = require("lodash.clonedeep");
+
+const getSemverExpression = source => {
+  return source.match(/pragma solidity(.*);/)[1].trim();
+};
+
+const getSemverExpressions = sources => {
+  return sources.map(source => getSemverExpression(source));
+};
+
+const validateSemverExpressions = semverExpressions => {
+  const { validRange } = semver;
+  for (const expression of semverExpressions) {
+    if (semver.validRange(expression) === null) {
+      const message = `Invalid semver expression (${expression}) found in` +
+        `one of your contract's imports.`;
+      throw new Error(message);
+    }
+  }
+};
+
+// takes an array of versions and an array of semver expressions
+// returns a version of the compiler or undefined if none can be found
+const findNewestSatisfyingVersion = ({ solcReleases, semverExpressions }) => {
+  // releases are ordered from newest to oldest
+  return solcReleases.find(version => {
+    return semverExpressions.every(expression =>
+      semver.satisfies(version, expression)
+    );
+  });
+};
+
+const throwCompilerVersionNotFound = ({ path, semverExpressions }) => {
+  const message =
+    `Could not find a single version of the Solidity compiler that ` +
+    `satisfies the following semver expressions obtained from your source ` +
+    `files' pragma statements: ${semverExpressions.join(" - ")}. ` +
+    `${OS.EOL}Please check the pragma statements for ${path} and its imports.`;
+  throw new Error(message);
+};
+
+const compileWithPragmaAnalysis = async ({ paths, options }) => {
+  const supplierOptions = {
+    events: options.events,
+    solcConfig: options.compilers.solc
+  };
+  const compilerSupplier = new CompilerSupplier(supplierOptions);
+  const { releases } = await compilerSupplier.getReleases();
+
+  // collect sources by the version of the Solidity compiler that they require
+  const versionsAndSources = {};
+  for (const path of paths) {
```

Comment:
Sorry, one more comment here I just realized: We should filter `paths` by `shouldIncludePath` or something similar, right?  Otherwise we risk attempting to run this pragma analyis on Vyper files.

Reply:
Oh I guess that is a good point. I missed that, good catch!